x/did/types/errors.go

```diff
@@ -0,0 +1,30 @@
+package types
+
+// DONTCOVER
+
+import (
+	"fmt"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// x/did module sentinel errors
+var (
+	ErrDIDExists = sdkerrors.Register(ModuleName, 101, "DID %v already exists")
+	ErrInvalidDID = sdkerrors.Register(ModuleName, 102, "Invalid DID %v")
+	ErrInvalidDIDDocument = sdkerrors.Register(ModuleName, 103, "Invalid DID Document: %v")
+	ErrDIDNotFound = sdkerrors.Register(ModuleName, 104, "DID %v not found")
+	ErrInvalidSignature = sdkerrors.Register(ModuleName, 105, "Invalid signature %v")
+	ErrInvalidVerificationMethodID = sdkerrors.Register(ModuleName, 106, "Invalid VerificationMethodID: %s")
+	ErrVerificationMethodIDNotFound = sdkerrors.Register(ModuleName, 107, "VerificationMethodID %v not found")
+	ErrSigVerificationFailed = sdkerrors.Register(ModuleName, 108, "DID signature verification was failed")
+	ErrInvalidSecp256k1PublicKey = sdkerrors.Register(ModuleName, 109, "Invalid Secp256k1 public key: %v")
+	ErrInvalidNetworkID = sdkerrors.Register(ModuleName, 110, "Invalid network ID: %s")
+	ErrInvalidDIDDocumentWithSeq = sdkerrors.Register(ModuleName, 111, "Invalid DIDDocumentWithSeq: %v")
+	ErrDIDDeactivated = sdkerrors.Register(ModuleName, 112, "DID was already deactivated: %v")
+	ErrVerificationMethodKeyTypeNotImplemented = sdkerrors.Register(ModuleName, 114, "Verification not implemented with key type: %v")
+
+)
+
+func Error(error *sdkerrors.Error, args ...interface{}) error {
+	return sdkerrors.New(error.Codespace(), error.ABCICode(), fmt.Sprintf(error.Error(), args))
+}
```

Comment:
I think it's better to use the error wrapping, instead of disassembling `sdkerrors.Error` for the string formatting.

For example, when you wanna wrap an `ErrInvalidDID` error with an additional message.
```go
sdkerrors.Wrap(ErrInvalidDID, "did:bitcoin:123125125")
```

Because Go doesn't have exceptions (+ err stacktrace), there have been a lot of discussions how to deal with errors nicely.
Finally, Go team decided to adopt the error wrapping concept: https://blog.golang.org/go1.13-errors#TOC_3. .
With this concept, we can check if the error `is`/`as` a certain error.

Furthermore,
Cosmos team wraps this concept one more time by defining the `sdkerror` package, which is what I don't like :). But, they recommend to use `sdkerrors.Wrap`, so that use can

Reply:
You can refer to https://github.com/medibloc/panacea-core/blob/d27d8818192eee72c192747a8a9e417346ae20ae/x/token/keeper/msg_server_token.go#L21

Reply:
@youngjoon-lee 

If it is wrapped, the error description is marked as two. 
In this case, `DID %v already exists` and `DID [did:asdfdsafsdafadsfdsafasdfdsf] already exists` are exposed together. 
I don't think it's a covering concept, but it can be seen as an error that occurs here. 

Please give me your opinion.

Reply:
Right. The current Cosmos `sdkerrors` sucks ㅜㅠ. The only one choice that we can choose is:
```go
var (
	ErrInvalidName = sdkerrors.Register(ModuleName, 1, "invalid name")  // without %v or %s
)
```
```go
name := "dummy"
return sdkerros.Wrap(ErrInvalidName, name)
```
Then, the composited error message might be:
```
invalid name: dummy
```

Reply:
In general, the recommended way is:
```go
err := errors.New("mem crash")
err = fmt.Errorf("failed to calculate your age: %w", err)
fmt.Println(err)
```
Then, the err message might be:
```
failed to calculate your age: mem crash
```
which is readable.

But, Cosmos uses their own special way...

Reply:
As we said before, modify it into a wrapping error.